client/rpc/src/state/state_full.rs

```diff
@@ -607,6 +607,29 @@ impl<BE, Block, Client> ChildStateBackend<Block, Client> for FullState<BE, Block
 				.map_err(client_err)))
 	}
 
+	fn storage_keys_paged(
+		&self,
+		block: Option<Block::Hash>,
+		storage_key: PrefixedStorageKey,
+		prefix: Option<StorageKey>,
+		count: u32,
+		start_key: Option<StorageKey>,
+	) -> FutureResult<Vec<StorageKey>> {
+		Box::new(result(
+			self.block_or_best(block)
+				.and_then(|block| {
+					let child_info = match ChildType::from_prefixed_key(&storage_key) {
+						Some((ChildType::ParentKeyId, storage_key)) => ChildInfo::new_default(storage_key),
+						None => return Err(sp_blockchain::Error::InvalidChildStorageKey),
+					};
+					self.client.child_storage_keys_iter(
+						&BlockId::Hash(block), child_info, prefix.as_ref(), start_key.as_ref(),
+					)
+				})
+				.map(|v| v.take(count as usize).collect())
```

Comment:
I thought this was interesting since it will always get the full state first, i.e. while saving on the responses over RPC, the work for retrieving 1 key vs 1024 keys are the same. Which is actually the time required to traverse the specified prefix.

Looked at the `state_getKeysPaged` and it seems the same, gets all then trims to count.

There must be a performance optimization here in both cases if I read it correctly, i.e. as there are more keys it will slow down.

Reply:
I copied paste the code, this snippet is a bit awkward (I read it the same way at first). `v` is an iterator (only does initial seek) and `take` will only call `next` on the iterator `count` time.
Agree the code give the wrong impression. Will rename `v` to `iter`.

Reply:
Cool. Glad it was my reading of the code - pre coffee this morning :)